### DISCUSSION TITLE
Inheritance in render.xml files

### DISCUSSION THREAD

**Discussion by Wonderer0** [osmandapp/OsmAnd#12237] [Maps Related]
I am experimenting writing a simple custom render.xml file to make hiking paths clearer by changing some of their colours and pathEffects, and by making the stroke width the same for all types but increasing with the zoom level (roughly sqrt(2) for each doubling):
```
<renderingStyle name="Experiment-1" depends="default" defaultColor="#ebe7e4" version="1">

  <renderingConstant name="anyPathMinZoom" value="12" />
  
  <renderingConstant name="pathPathEffect" value="4_2" />
  <renderingConstant name="footwayPathEffect" value="8_4" />
  <renderingConstant name="bridlewayPathEffect" value="12_6" />

  <renderingAttribute name="trackColor">
    <case nightMode="false" attrColorValue="#804000" /> 
  </renderingAttribute>

  <renderingAttribute name="pathColor">
    <case nightMode="false" attrColorValue="#C06000" /> 
  </renderingAttribute>

  <renderingAttribute name="footwayColor">
    <case nightMode="false" attrColorValue="#FF0000" /> 
  </renderingAttribute>

  <renderingAttribute name="bridlewayColor">
    <case nightMode="false" attrColorValue="#FF00FF" /> 
  </renderingAttribute>

  <line>
    <switch minzoom="2"> <!-- for new OpenGL engine -->
      <switch baseAppMode="pedestrian" moreDetailed="true" minzoom="$anyPathMinZoom" 
        tag="highway" strokeWidth="0.1">
        <case value="track" color_5="$trackColor" />
        <case value="path" color_5="$pathColor" pathEffect_5="$pathPathEffect" />
        <case value="footway" color_5="$footwayColor" pathEffect_5="$footwayPathEffect" />
        <case value="bridleway" color_5="$bridlewayColor" pathEffect_5="$bridlewayPathEffect" />
        <apply>
          <case minzoom="17" strokeWidth_4="5.6" strokeWidth_5="4" />
          <case minzoom="16" strokeWidth_4="4" strokeWidth_5="2.8" />
          <case minzoom="15" strokeWidth_4="2.8" strokeWidth_5="2" />
          <case minzoom="14" strokeWidth_4="2" strokeWidth_5="1.4" />
          <case minzoom="$anyPathMinZoom" strokeWidth_4="1.4" strokeWidth_5="1" />
        </apply>
      </switch>
    </switch>
  </line>
</renderingStyle>
```
It deliberately does NOT define the pathEffect for tracks, as I wanted that to be inherited from the default renderer, which changes the effect according to the path's grade.

It seems to be almost working as expected, except that all the tracks are solid lines.  Why aren't their path effects as defined in `default.render.xml` (lines 10295-10303) taking effect?

Could someone point to any documentation on how inheritance works for  render.xml files please.  In OsmAnd Docs [Map Rendering style - .render.xml](https://docs.osmand.net/en/main@latest/development/osmand-file-formats/osmand-rendering-style#map-style-sections) it just says that the header has a `depends` attribute that "inherits all properties from parent style".  What is meant by properties, is it just `<renderingProperty>`s?  It seems that `<renderingConstant>`s and `<renderingAttribute>`s aren't inherited, and so cause an error if used in the child file but not defined there again.  How are the `<line>` elements evaluated when they are in two or more files? 

The way in which `<case>` and other selectors are applied is described differently in the documentation at the top of `default.render.xml` and in [Map Rendering style - .render.xml](https://docs.osmand.net/en/main@latest/development/osmand-file-formats/osmand-rendering-style#map-style-sections).  The former suggests that they are evaluated sequentially, skipping over sections that do not apply when conditions fail.  However the latter says:

> To determine which attributes should be retrieved from rendering style, following search procedure is completed:
>
>   * Find the most inner <case> that corresponds to Search input parameters tag/value/zoom ...

which suggests that it works from the inside outwards somehow. 

Suppose there are two sibling `<case>` elements `c1` and `c2`, `c1` doesn't have any children, but `c2` has a child `c2.1`.  According to `default.render.xml` documentation `c1` would be evaluated first and if true then `c2` would be skipped, however according to OsmAnd Docs `c2.1` would be evaluated first because it is the inner most.

At first I assumed that parent and child files are evaluated sequentially, with later assignments in the child file replacing inherited ones (e.g. `pathColor`), but all other attributes remaining unchanged (e.g. track's `pathEffect`).  Unfortunately it doesn't seem to be that simple.  

It would be very straightforward to copy the lines that define track's `pathAffect`s in `default.render.xml` into the child file, but I am  am wondering what else need to be copied across because it isn't inherited automatically.  In that segment of the default renderer (10278-10480) there's lots of code to do with surface grades, winter roads, alpine hiking, tunnels, bridges, tolls etc.  Does all that need to be copied across to take effect?

P.S. Why aren't the paths rendered at all if `strokeWidth="0.1"` isn't given?  What's the difference between the it and `strokwWidth_4`  and `strokwWidth_5` ?  The former always seems to be larger than the latter, but why, and by how much should it be?  It doesn't seem to make any difference on the map.

**Comment by vshcherb** [comment]
Quite complicated part do you know
```
 <switch baseAppMode="pedestrian" moreDetailed="true" minzoom="$anyPathMinZoom" 
        tag="highway" strokeWidth="0.1">
```
Defines width only for pedestrian profile. I would start with something simple and start adding small chunks. Inheritance is pretty simple and it works as combining 2 styles into 1 file: New on top and based below.

**Reply by Wonderer0** [reply]
Yes, I want the paths to be more easy to see only for the pedestrian profile when "More details" is enabled, only when out walking or planning a walking route.

Glad to hear that inheritance is pretty simple.  When you say "New on top and based below" does that mean mine is processed first then the default one?  Wouldn't this mean that values assigned in my file (e.g. `strokeWidth_5`) would be replaced by ones in the default file?  Also why isn't `pathEffect_5` set correctly for tracks (depending upon their grade) by the default renderer, even though it is processed after my file, which doesn't set that value for tracks?

**Reply by Wonderer0** [reply]
Here is a much simpler example, all it does is change the strokeWidth and color of tracks:

```
<?xml version="1.0" encoding="utf-8"?>
<renderingStyle name="Experiment-2" depends="default" defaultColor="#ebe7e4" version="1">

  <renderingProperty attr="appMode" name="Rendering mode" description="Experiment-2"
    type="string" possibleValues="browse map,car,bicycle,pedestrian"/>

  <line>
    <switch minzoom="2">  <!-- for new OpenGL engine -->
      <switch strokeWidth="0.1">
        <case tag="highway" value="track" strokeWidth_4="8" strokeWidth_5="6" color_5="#804000" />
      </switch>
    </switch>
  </line>

</renderingStyle>
```

All the tracks are displayed as solid lines, even though the child renderer doesn't change their pathEffects.  It seems that the grade dependent pathEffects set in the default renderer aren't being inherited.

**Reply by vshcherb** [reply]
That's right behavior. In case you want to modify you need to embed 
```
<case tag="highway" value="track" strokeWidth_4="8" strokeWidth_5="6" color_5="#804000" >
   <case minzoom="4" strokeWidth_4="8" strokeWidth_5="6" color_5="#804000". ... >
</case>
```

If you put line next to each other they don't work separately only first one matched works, similar if / else if / else if ... in programming language

**Reply by Wonderer0** [reply]
I understand that after a `<case>` evaluates to `true` all the others after it at the same level in the same enclosing element are skipped, as in `<case value="track" ... />`, `<case value="path" ... />` in my original example.  The difference is that I wrapped the embedded `<case minzoom="17"  .../>`, `<case minzoom="16"  .../>` elements in an `<apply>` to group them together, which I thought the syntax required.

What I don't understand why in the simpler example the `trackEffect` property isn't inherited correctly from the default renderer.  It sets the `trackEeffect` to different values depending upon the track's grade, but they all appear as solid lines with the child renderer, even though it doesn't change the value of that property.
